File: test/VisualBasic/CodeCracker.Test/Performance/SealedAttributeTests.vb
Imports CodeCracker.VisualBasic.Performance
Imports Microsoft.CodeAnalysis.Testing
Imports Xunit

Namespace Performance
    Public Class SealedAttributeTests
        Inherits CodeFixVerifier(Of SealedAttributeAnalyzer, SealedAttributeCodeFixProvider)

        <Fact>
        Public Async Function ApplySealedWhenClassInheritsFromSystemAttributeClass() As Task
            Const test = "
Public Class MyAttribute
    Inherits System.Attribute
End Class"

            Dim expected = New DiagnosticResult(SealedAttributeAnalyzer.Id, Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(2, 14) _
                .WithMessage("Mark 'MyAttribute' as NotInheritable.")

            Await VerifyBasicDiagnosticAsync(test, expected)

        End Function

        <Fact>
        Public Async Function ApplySealedWhenClassInheritsIndirectlyFromSystemAttributeClass() As Task
            Const test = "
Public MustInherit Class MyAttribute
    Inherits System.Attribute
End Class

Public Class OtherAttribute
    Inherits MyAttribute
End Class"

            Dim expected = New DiagnosticResult(SealedAttributeAnalyzer.Id, Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(6, 14) _
                .WithMessage("Mark 'OtherAttribute' as NotInheritable.")

            Await VerifyBasicDiagnosticAsync(test, expected)
        End Function

        <Fact>
        Public Async Function NotApplySealedWhenClassThatInheritsFromSystemAttributeClassIsAbstract() As Task
            Const test = "
Public MustInherit Class MyAttribute
    Inherits System.Attribute
End Class"

            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function NotApplySealedWhenClassThatInheritsFromSystemAttributeClassIsNotInheritable() As Task
            Const test = "
Public NotInheritable Class MyAttribute
    Inherits System.Attribute
End Class"
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function NotApplyNotInheritableWhenIsStruct() As Task
            Const test = "
Public Structure MyStructure
End Structure"
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function NotApplyNotInheritableWhenIsInterface() As Task
            Const test = "
Public Interface ITest
End Interface"
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function WhenSealedModifierIsAppliedOnClass() As Task
            Const test = "
Public Class MyAttribute
    Inherits System.Attribute
End Class"

            Const fix = "
Public NotInheritable Class MyAttribute
    Inherits System.Attribute
End Class"

            Await VerifyBasicFixAsync(test, fix, 0)
        End Function

        <Fact>
        Public Async Function WhenSealedModifierIsAppliedOnClassFixAll() As Task
            Const test1 = "
Public Class MyAttribute1
    Inherits System.Attribute
End Class"
            Const fix1 = "
Public NotInheritable Class MyAttribute1
    Inherits System.Attribute
End Class"

            Const test2 = "
Public Class MyAttribute2
    Inherits System.Attribute
End Class"
            Const fix2 = "
Public NotInheritable Class MyAttribute2
    Inherits System.Attribute
End Class"

            Await VerifyBasicFixAllAsync(New String() {test1, test2}, New String() {fix1, fix2})
        End Function
    End Class
End Namespace

File: test/CSharp/CodeCracker.Test/Performance/SealedAttributeTests.cs
using CodeCracker.CSharp.Performance;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Testing;
using System.Threading.Tasks;
using Xunit;

namespace CodeCracker.Test.CSharp.Performance
{
    public class SealedAttributeTests : CodeFixVerifier<SealedAttributeAnalyzer, SealedAttributeCodeFixProvider>
    {
        [Fact]
        public async Task ApplySealedWhenClassInheritsFromSystemAttributeClass()
        {
            const string test = @"
                public class MyAttribute : System.Attribute
                {

                }";

            var expected = new DiagnosticResult(DiagnosticId.SealedAttribute.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(2, 30)
                .WithMessage("Mark 'MyAttribute' as sealed.");

            await VerifyCSharpDiagnosticAsync(test, expected);
        }

        [Fact]
        public async Task ApplySealedWhenClassInheritsIndirectlyFromSystemAttributeClass()
        {
            const string test = @"
                public abstract class MyAttribute : System.Attribute
                {

                }

                public class OtherAttribute : MyAttribute
                {

                }";

            var expected = new DiagnosticResult(DiagnosticId.SealedAttribute.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(7, 30)
                .WithMessage("Mark 'OtherAttribute' as sealed.");

            await VerifyCSharpDiagnosticAsync(test, expected);
        }

        [Fact]
        public async Task NotApplySealedWhenClassThatInheritsFromSystemAttributeClassIsAbstract()
        {
            const string test = @"
                public abstract class MyAttribute : System.Attribute
                {

                }";

            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task NotApplySealedWhenClassThatInheritsFromSystemAttributeClassIsSealed()
        {
            const string test = @"
                public sealed class MyAttribute : System.Attribute
                {

                }";

            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task NotApplySealedWhenIsStruct()
        {
            const string test = @"
                public struct MyStruct
                {

                }";

            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task NotApplySealedWhenIsInterface()
        {
            const string test = @"
                public interface ITest
                {

                    }";

            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task WhenSealedModifierIsAppliedOnClass()
        {
            const string source = @"
                public class MyAttribute : System.Attribute
                {
                }";

            const string fixtest = @"
                public sealed class MyAttribute : System.Attribute
                {
                }";

            await VerifyCSharpFixAsync(source, fixtest, 0);
        }

        [Fact]
        public async Task WhenSealedModifierIsAppliedOnClassFixAll()
        {
            const string source1 = @"
                public class MyAttribute1 : System.Attribute
                {
                }
                public class MyAttribute3 : System.Attribute
                {
                }";
            const string fixtest1 = @"
                public sealed class MyAttribute1 : System.Attribute
                {
                }
                public sealed class MyAttribute3 : System.Attribute
                {
                }";

            const string source2 = @"
                public class MyAttribute2 : System.Attribute
                {
                }";
            const string fixtest2 = @"
                public sealed class MyAttribute2 : System.Attribute
                {
                }";

            await VerifyCSharpFixAllAsync(new string[] { source1, source2 }, new string[] { fixtest1, fixtest2 });
        }
    }
}